### PR TITLE
feat(ape-staking): adds ape staking feature

### PR DESCRIPTION
**Pull request by WalidOfNow**
## Security Checklist

- [ ] 1. Re-Entrancy
- [ ] 2. Arithmetic Over/Under Flows
- [ ] 3. Unexpected Ether
- [ ] 4. Delegatecall
- [ ] 5. Default Visibilities
- [ ] 6. Entropy Illusion
- [ ] 7. External Contract Referencing
- [ ] 8. Short Address/Parameter Attack (off chain)
- [ ] 9. Unchecked CALL Return Values
- [ ] 10. Race Conditions / Front Running
- [ ] 11. Denial Of Service (DOS)
- [ ] 12. Block Timestamp Manipulation
- [ ] 13. Constructors with Care
- [ ] 14. Uninitialized Storage Pointers
- [ ] 15. Floating Points and Precision
- [ ] 16. Tx.Origin Authentication
- [ ] 17. Address.isContract Re-Entrancy via Constructor

### ⚠️ NOTES ⚠️

Make sure to think about each of these exploits in this PR.

### CODE REVIEW

File: contracts/protocol/tokenization/NTokenBAYC.sol [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        _apeCoinStaking = ApeCoinStaking(apeCoinStaking);
+    }
+
+    /**
+     * @notice Deposit ApeCoin to the BAYC Pool
+     * @param _nfts Array of SingleNft structs
+     * @dev Commits 1 or more BAYC NFTs, each with an ApeCoin amount to the BAYC pool.\
+     * Each BAYC committed must attach an ApeCoin amount >= 1 ApeCoin and <= the BAYC pool cap amount.
+     */
+    function depositBAYC(ApeCoinStaking.SingleNft[] calldata _nfts) external {
```
Comment (GopherJ) [inline contracts/protocol/tokenization/NTokenBAYC.sol:33]: there is no re-entrancy guard

**Comment by yrong** [comment]
@WalidOfNow seems mixed with a lot of format changes so difficult to review, could we separate it from this pr?

**Comment by zhoujia6139** [comment]
Can we extract the repeated code logic between NTokenBAYC and NTokenMAYC to a NTokenApeStaking.

File: contracts/protocol/tokenization/NTokenBAYC.sol [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+                    tokenId: tokenIds[index],
+                    amount: stakedAmount
+                });
+                counter++;
+            }
+        }
+
+        if (counter > 0) {
+            assembly {
+                mstore(nfts, sub(mload(nfts), sub(tokenIdLength, counter)))
```
Comment (zhoujia6139) [inline contracts/protocol/tokenization/NTokenBAYC.sol:162]: Can we simplify this to
`mstore(nfts, counter))`

**Comment by WalidOfNow** [comment]
@yrong you only need to look at the NToken contracts. you can ignore other contracts as they are just formatted.

File: contracts/protocol/tokenization/NTokenApeStaking.sol [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+    function depositApeCoin(ApeCoinStaking.SingleNft[] calldata _nfts)
+        external
+    {
+        for (uint256 index = 0; index < _nfts.length; index++) {
+            require(
+                ownerOf(_nfts[index].tokenId) == msg.sender,
+                "NToken: not owner of token"
+            );
+
+            _apeCoinStaking.apeCoin().transferFrom(
```
Comment (zhoujia6139) [inline contracts/protocol/tokenization/NTokenApeStaking.sol:64]: can we sum the ape token amount and transfer once?

File: contracts/protocol/tokenization/NTokenApeStaking.sol [hunk 4]
```diff
@@ -0,0 +1,193 @@
...
+        }
+
+        _withdrawApeCoin(_nfts, _recipient);
+    }
+
+    /**
+     * @notice Overrides the transferOnLiquidation from NToken to withdraw all staked and pending rewards before transfer the asset on liquidation
+     */
+    function transferOnLiquidation(address from, address to, uint256 tokenId)
+        public
```
Comment (zhoujia6139) [inline contracts/protocol/tokenization/NTokenApeStaking.sol:112]: can we change public to external?

File: contracts/protocol/tokenization/NTokenApeStaking.sol [hunk 5]
```diff
@@ -0,0 +1,193 @@
...
+     * @notice Withdraw staked ApeCoin from the BAYC pool.  If withdraw is total staked amount, performs an automatic claim.
+     * @param _nfts Array of BAYC NFT's with staked amounts
+     * @param _recipient Address to send withdraw amount and claim to
+     */
+    function withdrawApeCoin(
+        ApeCoinStaking.SingleNft[] calldata _nfts,
+        address _recipient
+    ) external {
+        for (uint256 index = 0; index < _nfts.length; index++) {
+            require(ownerOf(_nfts[index].tokenId) == msg.sender);
```
Comment (zhoujia6139) [inline contracts/protocol/tokenization/NTokenApeStaking.sol:102]: can we add error message here "NToken: not owner of token" ?

**Review by GopherJ — APPROVED** [review]

**Pull request merged**